document-portal/file-transfer.c

```diff
@@ -0,0 +1,544 @@
+#include "document-portal-dbus.h"
+#include "document-enums.h"
+#include "document-portal.h"
+#include "document-portal-fuse.h"
+
+static XdpDbusFileTransfer *file_transfer;
+
+typedef struct
+{
+  char *path;
+  int parent_dev;
+  int parent_ino;
+} ExportedFile;
+
+static void
+exported_file_free (gpointer data)
+{
+  ExportedFile *file = data;
+
+  g_free (file->path);
+  g_free (file);
+}
+
+typedef struct
+{
+  GObject object;
+  GMutex mutex;
+
+  GPtrArray *files;
+  gboolean writable;
+  gboolean autostop;
+  char *key;
+  char *sender;
+  XdpAppInfo *app_info;
+} FileTransfer;
+
+typedef struct
+{
+  GObjectClass parent_class;
+} FileTransferClass;
+
+static GType file_transfer_get_type (void);
+
+G_DEFINE_TYPE (FileTransfer, file_transfer, G_TYPE_OBJECT)
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (FileTransfer, g_object_unref);
+
+static void
+file_transfer_init (FileTransfer *transfer)
+{
+  g_mutex_init (&transfer->mutex);
+}
+
+static void
+file_transfer_finalize (GObject *object)
+{
+  FileTransfer *transfer = (FileTransfer *)object;
+
+  g_mutex_clear (&transfer->mutex);
+  xdp_app_info_unref (transfer->app_info);
+  g_ptr_array_unref (transfer->files);
+  g_free (transfer->key);
+  g_free (transfer->sender);
+
+  G_OBJECT_CLASS (file_transfer_parent_class)->finalize (object);
+}
+
+static void
+file_transfer_class_init (FileTransferClass *class)
+{
+  G_OBJECT_CLASS (class)->finalize = file_transfer_finalize;
+}
+
+static inline void
+auto_unlock_unref_helper (FileTransfer **transfer)
+{
+  if (!*transfer)
+    return;
+
+  g_mutex_unlock (&(*transfer)->mutex);
+  g_object_unref (*transfer);
+}
+
+static inline FileTransfer *
+auto_lock_helper (FileTransfer *transfer)
+{
+  if (transfer)
+    g_mutex_lock (&transfer->mutex);
+  return transfer;
+}
+
+#define TRANSFER_AUTOLOCK_UNREF(transfer) \
+  G_GNUC_UNUSED __attribute__((cleanup (auto_unlock_unref_helper))) \
+  FileTransfer * G_PASTE (auto_unlock_unref, __LINE__) = \
+    auto_lock_helper (transfer);
+
+G_LOCK_DEFINE (transfers);
+static GHashTable *transfers;
+
+static FileTransfer *
+lookup_transfer (const char *key)
+{
+  FileTransfer *transfer;
+
+  G_LOCK (transfers);
+  transfer = (FileTransfer *)g_hash_table_lookup (transfers, key);
+  if (transfer)
+    g_object_ref (transfer);
+  G_UNLOCK (transfers);
+
+  return transfer;
+}
+
+static FileTransfer *
+file_transfer_start (XdpAppInfo *app_info,
+                     const char *sender,
+                     gboolean    writable,
+                     gboolean    autostop)
+{
+  FileTransfer *transfer;
+
+  transfer = g_object_new (file_transfer_get_type (), NULL);
+
+  transfer->app_info = xdp_app_info_ref (app_info);
+  transfer->sender = g_strdup (sender);
+  transfer->writable = writable;
+  transfer->autostop = autostop;
+  transfer->files = g_ptr_array_new_with_free_func (exported_file_free);
+
+  G_LOCK (transfers);
+  do {
+    g_free (transfer->key);
+    transfer->key = g_strdup_printf ("%u", g_random_int ());
```

Comment:
I don't think a 32bit random value is enough to protect against a brute force attack. Calling retrieve_files() 4 billion times seem vaguely realistic, while bumping it to 64bit would make it unthinkable.